contracts/external/dao-migrator/src/contract.rs

```diff
@@ -0,0 +1,388 @@
+use std::env;
+
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    to_binary, Addr, Binary, CosmosMsg, Deps, DepsMut, Env, MessageInfo, Reply, Response,
+    StdResult, SubMsg, WasmMsg,
+};
+use cw2::set_contract_version;
+use dao_core::{query::SubDao, state::ProposalModule};
+use dao_interface::ModuleInstantiateCallback;
+
+use crate::{
+    error::ContractError,
+    msg::{ExecuteMsg, InstantiateMsg, MigrateV1ToV2, QueryMsg},
+    state::{CORE_ADDR, MODULES_ADDRS, TEST_STATE},
+    types::{
+        CodeIdPair, MigrationMsgs, MigrationParams, ModulesAddrs, TestState, V1CodeIds, V2CodeIds,
+    },
+    utils::state_queries::{
+        query_proposal_count_v1, query_proposal_count_v2, query_proposal_v1, query_proposal_v2,
+        query_single_voting_power_v1, query_single_voting_power_v2, query_total_voting_power_v1,
+        query_total_voting_power_v2,
+    },
+};
+
+pub(crate) const CONTRACT_NAME: &str = "crates.io:dao-migrator";
+pub(crate) const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+pub(crate) const V1_V2_REPLY_ID: u64 = 1;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    CORE_ADDR.save(deps.storage, &info.sender)?;
+
+    Ok(
+        Response::default().set_data(to_binary(&ModuleInstantiateCallback {
+            msgs: vec![WasmMsg::Execute {
+                contract_addr: env.contract.address.to_string(),
+                msg: to_binary(&MigrateV1ToV2 {
+                    sub_daos: msg.sub_daos,
+                    migration_params: msg.migration_params,
+                    v1_code_ids: msg.v1_code_ids,
+                    v2_code_ids: msg.v2_code_ids,
+                })?,
+                funds: vec![],
+            }
+            .into()],
+        })?),
+    )
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    execute_migration_v1_v2(
+        deps,
+        env,
+        info,
+        msg.sub_daos,
+        msg.migration_params,
+        msg.v1_code_ids,
+        msg.v2_code_ids,
+    )
+}
+
+fn execute_migration_v1_v2(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    sub_daos: Vec<SubDao>,
+    migration_params: MigrationParams,
+    v1_code_ids: V1CodeIds,
+    v2_code_ids: V2CodeIds,
+) -> Result<Response, ContractError> {
```

Comment:
could we check `info.sender == CORE_ADDR.load(deps.storage)` here? so long as this is atomic, this shouldn't matter, but having the check makes this safer in the event that we mess up atomicity.